tests/integration/sharding_tests/test_mongos.py

```diff
@@ -0,0 +1,154 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
+import pytest
+from pymongo.errors import OperationFailure
+from pytest_operator.plugin import OpsTest
+
+from .helpers import count_users, generate_mongodb_client, get_username_password
+
+MONGOS_HOST_APP_NAME = "application"
+SHARD_ONE_APP_NAME = "shard-one"
+MONGOS_APP_NAME = "mongos"
+CONFIG_SERVER_APP_NAME = "config-server-one"
+SHARD_REL_NAME = "sharding"
+CLUSTER_REL_NAME = "cluster"
+CONFIG_SERVER_REL_NAME = "config-server"
+MONGODB_KEYFILE_PATH = "/var/snap/charmed-mongodb/current/etc/mongod/keyFile"
+TIMEOUT = 10 * 60
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest, mongos_host_application_charm) -> None:
+    """Build and deploy a sharded cluster."""
+    mongodb_charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(
+        mongodb_charm,
+        num_units=1,
+        config={"role": "config-server"},
+        application_name=CONFIG_SERVER_APP_NAME,
+    )
+    await ops_test.model.deploy(
+        mongodb_charm, num_units=1, config={"role": "shard"}, application_name=SHARD_ONE_APP_NAME
+    )
+
+    await ops_test.model.deploy(
+        MONGOS_APP_NAME,
+        channel="6/edge",
+        revision=3,
+    )
+
+    # TODO: Future PR, once data integrator works with mongos charm deploy that charm instead of
+    # packing and deploying the charm in the application dir.
+    await ops_test.model.deploy(
+        mongos_host_application_charm, application_name=MONGOS_HOST_APP_NAME
+    )
+
+    await ops_test.model.wait_for_idle(
+        apps=[CONFIG_SERVER_APP_NAME, SHARD_ONE_APP_NAME, MONGOS_HOST_APP_NAME],
+        idle_period=20,
+        raise_on_blocked=False,  # cluster components are blocked waiting for integration.
+        timeout=TIMEOUT,
+        raise_on_error=False,
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_form_cluster(ops_test: OpsTest) -> None:
```

Comment:
Could we be more explicit regarding the test name, to reflect that we're interested in user counts?